jax/random.py

```diff
@@ -1224,8 +1223,36 @@ def logistic(key, shape=(), dtype=onp.float64):
 
 @partial(jit, static_argnums=(1, 2))
 def _logistic(key, shape, dtype):
+  # Mathematically, we should draw a uniform variable x in [0, 1], and
+  # compute the logistic random variable using log(x / (1 - x)).
+  # An issue arises when implementing this directly via uniform(), because
+  #  1) uniform() generates values in the semi-closed interval [0, 1)
+  #  2) when x=0, the output is -inf
+  #  3) uniform() only randomizes the mantissa, meaning that for float32 input
+  #     the cardinality is only about 2^23 which means these infinities are quite
+  #     common (about 1 in 8 million)
+  # We can address this by generating integers I in (0, N), then transform
+  # using log(I / (N - I)). Note that the endpoints here are exclusive, so
+  # infinities are avoided at either end.
+  # But what should we choose for N? It's most efficient to generate integers
+  # in a range [0, 2^M) for integer M. We can add 1 to get (0, 2^M + 1).
+  # We cannot use M = nbits because 2^nbits overflows, so we use M = nbits - 1
+  # to get the highest possible cardinality in the most efficient manner.
+  # (Note: we avoid randint because it generates 2x more bits than necessary).
   _check_shape("logistic", shape)
-  return logit(uniform(key, shape, dtype))
+  nbits = np.finfo(dtype).bits
+  if nbits not in (32, 64):
```

Comment:
You should support 16-bit types also. A valid strategy for doing so it to generate 32-bit values and cast.

Reply:
``_random_bits`` only supports 32 and 64 bit https://github.com/google/jax/blob/28bc4b759ee113815bfacb99546e58a91891b3f2/jax/random.py#L273-L274
As do most other routines in this package.

Perhaps we should add a TODO to support 16-bit broadly within ``random()``?

Reply:
Yes, that seems like a bug if any of these don't support 16 bit floats.

But note you don't need 16 bit support in `_random_bits` necessarily to make this case work (in fact, it's probably preferable to use 32-bit math here.)

Reply:
Since the existing ``logistic()`` function only supports 32/64 bit, I think it makes sense to not expand the scope of this bug fix to adding more dtypes.

I opened #3062 to track this enhancement.